### PR TITLE
L89: C-core: Remove grpc_insecure_channel_create and grpc_server_add_insecure_http2_port from Core Surface API

### PR DESCRIPTION
**Pull request by yihuazhang**

### CODE REVIEW

File: L87-core-remove-grpc-insecure-channel-creation-api
Comment (sanjaypujare) [inline L87-core-remove-grpc-insecure-channel-creation-api:110]: It might be useful to note that this will have negligible or zero impact because this is a core API and not an end user API so users are likely to be affected by this removal (assuming true).

Comment (morgwai) [reply L87-core-remove-grpc-insecure-channel-creation-api:110]: i strongly agree with @sanjaypujare : when I first saw the thread on the mailing-list, I almost made my pants dirty and calmed down only after I saw his above comment.

Comment (yihuazhang) [reply L87-core-remove-grpc-insecure-channel-creation-api:110]: Thanks! I updated the text to reflect it.

Comment (sanjaypujare) [reply L87-core-remove-grpc-insecure-channel-creation-api:110]: The new text does acknowledge impact on users who "directly interact with Core surface APIs (e.g., by calling insecure channel creation API's)". Assuming such users exist, won't they benefit from deprecation first and then removal of the APIs?

Comment (yihuazhang) [reply L87-core-remove-grpc-insecure-channel-creation-api:110]: I do not think we need to keep backward-compatibility for C-core API users. Here is a quote from @markdroth  - "The C-core API is not a public API, and we don't offer a backward-compatibility guarantee for it. We are required to publish a gRFC for C-core API changes, but we are not required to make non-breaking changes"

Comment (markdroth) [reply L87-core-remove-grpc-insecure-channel-creation-api]: That's correct, we do not promise backward compatibility for the C-core API.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:1]: This filename needs a .md suffix so that github will format it correctly.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:22]: Please change the wording here to make it clear that we are actually making these changes.  Instead of "could" or "would", please use "will".

We should also explain why we're doing this.  Key points:
- The current approach requires different API entry points for secure vs. insecure channels and servers, and it requires a lot of conditional compilation that complicates our build system.
- The main reason for supporting insecure builds is because there are some users that don't need security but really need to minimize binary size (e.g., embedded systems), so they need a way to build without depending on the SSL libraries.
- The current approach was decided upon early in our development, before we had the credential APIs, and it was never updated when the credential APIs were added.  However, the credential APIs provide a much cleaner way to allow building gRPC without the SSL dependency.

To solve this problem, we are going to move to an approach where insecure channels and servers are handled the same way as secure channels and servers.  It will no longer be possible to build gRPC without support for credentials or to create a channel or server without specifying credentials, but we will provide an insecure credentials type to be used in cases where security is not needed.  The insecure credentials type will not depend on SSL libraries and will be included in all gRPC builds.  Other credential types that do depend on SSL libraries will be available in secure builds only, just as they are today.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:81]: I think these functions should stay in grpc_posix.h. They're not supported on non-POSIX systems.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:83]: Please remove the apostrophe in "API's".

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:106]: Let's get rid of this `reserved` parameter while we're changing this. I don't think we need it.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api:118]: s/bringin/bring in/

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api]: Done.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api]: Done.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api]: Done.

File: L89-core-remove-grpc-insecure-channel-creation-api
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api]: Thanks for adding more contexts. I added them to the new text.

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api.md:15]: Looks like you need to add a space before "related".

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (markdroth) [inline L89-core-remove-grpc-insecure-channel-creation-api.md:41]: Need a space before "related" here too.

**Review by markdroth — APPROVED** [review]
Just a couple of remaining formatting nits; otherwise, this looks great!

@jtattermusch and @ctiller may also want to review.

As per gRFC comment policy, we'll need to wait until November 15th to merge this.

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (jtattermusch) [inline L89-core-remove-grpc-insecure-channel-creation-api.md:17]: Listing the API changes you're planning is definitely a must, but since a big portion of the design discussions were also about the impact on the build targets, it probably makes sense to add a few sentences that describe the impact on existing build targets (e.g. `grpc_unsecure` and `grpc` which play the most important role here).
- let's mention the planned impact on the existing public bazel targets
- let's also mention the impact on the non-bazel targets (which are mostly represented by our cmake targets `grpc` and `grpc_unsecure`.

From you implementation PR, it looks like the `//:grpc` and `//:grpc_unsecure` targets will be kept and `//:grpc_unsecure` won't have a dependency on boringssl.
The cmake targets: it looks like we will still have `grpc` (with unconditional dependency on boringssl/openssl) and `grpc_unsecure` (NO  dependency on boringssl/openssl, also unconditionally).

If the above is correct, the proposal LGTM to me after adding the summary to the abstract.

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (jtattermusch) [inline L89-core-remove-grpc-insecure-channel-creation-api.md:126]: I'm curious if the issue https://github.com/grpc/grpc/issues/27733 (which I mentioned in https://github.com/grpc/grpc/pull/25586#issuecomment-944154028) will be resolved by the changes in the implementation of this proposal. If so, that's very positive and it should be mentioned here.

If not, perhaps we should mention it in the "Open issues" section, since in our own codebase, we're not always honoring the mutual exclusivity of  grpc and grpc_unsecure and we do have targets (e.g. some tests) that transitively depend on both `grpc` and `grpc_unsecure`.

**Review by jtattermusch — COMMENTED** [review]
LGTM after addressing my comments.

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api.md:19]: > From you implementation PR, it looks like the //:grpc and //:grpc_unsecure targets will be kept and //:grpc_unsecure won't have a dependency on boringssl.
The cmake targets: it looks like we will still have grpc (with unconditional dependency on boringssl/openssl) and grpc_unsecure (NO dependency on boringssl/openssl, also unconditionally).

That's correct, Jan. I also updated the text to make it more clear. Thanks for pointing it out.

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (yihuazhang) [reply L89-core-remove-grpc-insecure-channel-creation-api.md:135]: Ah, I do not think there is a need to keep `_nosec` end2end tests any more as we removed all code related to insecure channel creation (e.g., `grpc_insecure_channel_create`) API in this gRFC, that are exercised by `_nosec` tests.

Comment (jtattermusch) [reply L89-core-remove-grpc-insecure-channel-creation-api.md:135]: I think that's a good point. If we no longer need the `_nosec` tests after merging this PR (do other reviewers agree with this statement?), we should definitely get rid of those tests in a followup PR. There's actually quite a bit of complexity in how the _SEC vs _NOSEC end2end tests get configured and run, so this cleanup will be very welcome.

**Review by jtattermusch — APPROVED** [review]

File: L89-core-remove-grpc-insecure-channel-creation-api.md
Comment (markdroth) [reply L89-core-remove-grpc-insecure-channel-creation-api.md:135]: Yes, I agree, we no longer need the `_nosec` tests.  I'm tempted to say that we should just remove those as part of Yihua's existing PR, but the PR is already quite large and invasive as it is, so I guess let's do that as a separate step.

**Comment by yihuazhang** [comment]
@asraa, @yanavlasov I noticed `InsecureChannelCredentials` is used in envoy code base (e.g., [here](https://github.com/envoyproxy/envoy/blob/main/source/common/grpc/google_grpc_creds_impl.cc#L117-L119)), and I am wondering if the change in this gRFC (i.e., fail RPC if call credentials are sent over insecure channels) will break envoy tests. Any information will be appreciated.

**Comment by yanavlasov** [comment]
> @asraa, @yanavlasov I noticed `InsecureChannelCredentials` is used in envoy code base (e.g., [here](https://github.com/envoyproxy/envoy/blob/main/source/common/grpc/google_grpc_creds_impl.cc#L117-L119)), and I am wondering if the change in this gRFC (i.e., fail RPC if call credentials are sent over insecure channels) will break any envoy tests. Any information will be appreciated.

I do not know for sure. There are some tests that use credentials here https://github.com/envoyproxy/envoy/blob/ae56bb42102df22401c5c0df9d3f8341b18acdfc/test/common/grpc/grpc_client_integration_test.cc#L450

If there is a simple way to do it, we can do dry-run and see how many tests will break in Envoy. But we can also just fix them when we pull a new version of gRPC.

**Comment by yihuazhang** [comment]
> But we can also just fix them when we pull a new version of gRPC.
SG. PLMK if you find tests broken when importing the new version of gRPC, I am happy to help investigating.

**Comment by yihuazhang** [comment]
I believe we can merge this proposal now.

**Pull request merged**
